### PR TITLE
Issue #3054 Add command to start and stop systemtray in background

### LINKED ISSUE #3054: A way to start system tray without starting the cluster
I would like to be able to start the system tray without starting the cluster. I have tried `minishift services systemtray` which starts the systemtray but it is not permanent.

### PR DESCRIPTION
**Pull request by anjannath**
Fixes #3054 #3062 

Steps to test the pull request

Use these binaries for [windows](https://5829-62728051-gh.circle-artifacts.com/0/ms/windows-amd64/systemtray/minishift.exe) and [darwin](https://5829-62728051-gh.circle-artifacts.com/0/ms/darwin-amd64/systemtray/minishift)

  1. `minishift services start systemtray` ==> Starts the tray in background
  2. `minishift start` ==> shouldn't launch a new tray, as one is already running
  3. `minishift service stop systemtray` ==> kills the tray process

Going forward we can daemonize this command and make the tray launch at os start.

### CODE REVIEW

File: cmd/minishift/cmd/services/services.go [hunk 1]
```diff
@@ -27,5 +27,4 @@ var ServicesCmd = &cobra.Command{
 	Run: func(cmd *cobra.Command, args []string) {
 		cmd.Help()
 	},
-	Hidden: true,
```
Comment (praveenkumar) [inline cmd/minishift/cmd/services/services.go]: @anjannath so now we are exposing this command to users?

File: cmd/minishift/cmd/services/systray.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+// systrayCmd represents the systemtray command
+var systrayServiceCmd = &cobra.Command{
+	Use:   "systray",
+	Short: "Run minishift systemtray",
+	Long:  "Run a systemtray in the notification area of top bar/start menu",
+	Run: func(cmd *cobra.Command, args []string) {
+		systray.Run(systemtray.OnReady, systemtray.OnExit)
+	},
+	Hidden: true,
```
Comment (praveenkumar) [inline cmd/minishift/cmd/services/systray.go]: @anjannath if we are exposing the `service` command then why this subcommand we are hiding?

Comment (anjannath) [reply cmd/minishift/cmd/services/systray.go]: @praveenkumar This command `minishift services systray` do not put the tray to background, we call this command from `minishift services systemtray` after setting required flags for the process to make it a background process. So we don't need to expose the `systray` command to users.

File: cmd/minishift/cmd/services/services.go (hunk 1 above)
Comment (anjannath) [reply cmd/minishift/cmd/services/services.go]: Yes

**Comment by agajdosi** [comment]
Hi @anjannath, trying this on Windows and I am not able to start the tray from the CLI:
```
PS C:\Users\cdk> minishift services systemtray
Manages services used by Minishift.

Usage:
  minishift services SUBCOMMAND [flags]

Flags:
  -h, --help   help for services

Global Flags:
      --alsologtostderr                  log to standard error as well as files
      --httptest.serve string            if non-empty, httptest.NewServer serves on this address and blocks
      --log_backtrace_at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                   If non-empty, write log files in this directory
      --logtostderr                      log to standard error instead of files
      --profile string                   Profile name (default "minishift")
      --show-libmachine-logs             Show logs from libmachine.
      --stderrthreshold severity         logs at or above this threshold go to stderr (default 2)
      --test.bench regexp                run only benchmarks matching regexp
      --test.benchmem                    print memory allocations for benchmarks
      --test.benchtime d                 run each benchmark for duration d (default 1s)
      --test.blockprofile file           write a goroutine blocking profile to file
      --test.blockprofilerate rate       set blocking profile rate (see runtime.SetBlockProfileRate) (default 1)
      --test.count n                     run tests and benchmarks n times (default 1)
      --test.coverprofile file           write a coverage profile to file
      --test.cpu list                    comma-separated list of cpu counts to run each test with
      --test.cpuprofile file             write a cpu profile to file
      --test.failfast                    do not start new tests after the first test failure
      --test.list regexp                 list tests, examples, and benchmarks matching regexp then exit
      --test.memprofile file             write an allocation profile to file
      --test.memprofilerate rate         set memory allocation profiling rate (see runtime.MemProfileRate)
      --test.mutexprofile string         write a mutex contention profile to the named file after execution
      --test.mutexprofilefraction int    if >= 0, calls runtime.SetMutexProfileFraction() (default 1)
      --test.outputdir dir               write profiles to dir
      --test.parallel n                  run at most n tests in parallel (default 4)
      --test.run regexp                  run only tests and examples matching regexp
      --test.short                       run smaller test suite to save time
      --test.testlogfile file            write test action log to file (for use only by cmd/go)
      --test.timeout d                   panic test binary after duration d (default 0, timeout disabled) (default 0s)
      --test.trace file                  write an execution trace to file
      --test.v                           verbose: print additional output
  -v, --v Level                          log level for V logs. Level varies from 1 to 5 (default 1).
      --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
PS C:\Users\cdk>
```

**Comment by agajdosi** [comment]
Another suggestion is whether we could add `service` as alias for `services` subcommand. What do you think about it?

**Comment by anjannath** [comment]
@agajdosi Sure, i'll add alias `service`, also could you try these binaries for [windows](https://5815-62728051-gh.circle-artifacts.com/0/ms/windows-amd64/systemtray/minishift.exe) and [darwin](https://5815-62728051-gh.circle-artifacts.com/0/ms/darwin-amd64/systemtray/minishift)

**Comment by agajdosi** [comment]
@anjannath As mentioned above the command does not open anything on Win10 for me ^^

**Comment by agajdosi** [comment]
Used wrong binaries, need to get ones from CircleCI, will check it on Wednesday. (Tomorrow PTO :( )

**Comment by agajdosi** [comment]
Used wrong binaries, need to get ones from CircleCI, will check it on Wednesday. (Tomorrow PTO :( )

**Comment by anjannath** [comment]
Yes, that is correct.

File: cmd/minishift/cmd/services/systray.go (hunk 2 above)
Comment (anjannath) [reply cmd/minishift/cmd/services/systray.go]: @praveenkumar Yes that is correct.

**Comment by anjannath** [comment]
@agajdosi Use the binaries i linked in https://github.com/minishift/minishift/pull/3118#issuecomment-454020279

File: cmd/minishift/cmd/services/services.go (hunk 1 above)
Comment (gbraad) [reply cmd/minishift/cmd/services/services.go]: did we expose the Services namespace with a description... I remember I add one, but is it shown?

Comment (anjannath) [reply cmd/minishift/cmd/services/services.go]: @gbraad It shows the description

File: cmd/minishift/cmd/services/systemtray.go [hunk 3]
```diff
@@ -32,9 +32,9 @@ var systemtrayServiceCmd = &cobra.Command{
 	Short: "Run minishift systemtray",
 	Long:  "Run a systemtray in the notification area of top bar/start menu",
 	Run: func(cmd *cobra.Command, args []string) {
-		systray.Run(systemtray.OnReady, systemtray.OnExit)
+		minishiftTray := systemtray.NewMinishiftTray(minishiftConfig.AllInstancesConfig)
+		minishiftTray.EnsureRunning()
```
Comment (gbraad) [inline cmd/minishift/cmd/services/systemtray.go]: :+1

**Comment by LalatenduMohanty** [comment]
Also `make systemtray` is failing for me on macOS

```
$ make systemtray
go install -tags "containers_image_openpgp containers_image_ostree_stub exclude_graphdriver_btrfs exclude_graphdriver_devicemapper exclude_graphdriver_overlay containers_image_storage_stub" -pkgdir=/Users/lmohanty/code/gocode/src/github.com/minishift/minishift/out/bindata -ldflags="-X github.com/minishift/minishift/pkg/version.minishiftVersion=1.29.0 -X github.com/minishift/minishift/pkg/version.centOsIsoVersion=v1.13.0 -X github.com/minishift/minishift/pkg/version.openshiftVersion=v3.11.0 -X github.com/minishift/minishift/pkg/version.commitSha=0f6e0d14" ./cmd/minishift
# github.com/minishift/minishift/vendor/github.com/anjannath/systray
systray_darwin.m:82:15: warning: 'setImage:' is deprecated: first deprecated in macOS 10.14 - Use the receiver's button.image instead [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSStatusItem.h:85:39: note: property 'image' is declared deprecated here
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSStatusItem.h:85:39: note: 'setImage:' has been explicitly marked deprecated here
systray_darwin.m:86:15: warning: 'setTitle:' is deprecated: first deprecated in macOS 10.14 - Use the receiver's button.title instead [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSStatusItem.h:83:38: note: property 'title' is declared deprecated here
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSStatusItem.h:83:38: note: 'setTitle:' has been explicitly marked deprecated here
systray_darwin.m:90:15: warning: 'setToolTip:' is deprecated: first deprecated in macOS 10.14 - Use the receiver's button.toolTip instead [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSStatusItem.h:89:38: note: property 'toolTip' is declared deprecated here
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSStatusItem.h:89:38: note: 'setToolTip:' has been explicitly marked deprecated here
systray_darwin.m:128:24: warning: 'NSOnState' is deprecated: first deprecated in macOS 10.14 [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSCell.h:80:34: note: 'NSOnState' has been explicitly marked deprecated here
systray_darwin.m:130:24: warning: 'NSOffState' is deprecated: first deprecated in macOS 10.14 [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSCell.h:79:34: note: 'NSOffState' has been explicitly marked deprecated here
systray_darwin.m:158:24: warning: 'NSOnState' is deprecated: first deprecated in macOS 10.14 [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSCell.h:80:34: note: 'NSOnState' has been explicitly marked deprecated here
systray_darwin.m:160:24: warning: 'NSOffState' is deprecated: first deprecated in macOS 10.14 [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/AppKit.framework/Headers/NSCell.h:79:34: note: 'NSOffState' has been explicitly marked deprecated here
# github.com/minishift/minishift/cmd/minishift
/usr/local/go/pkg/tool/darwin_amd64/link: /usr/local/go/pkg/tool/darwin_amd64/link: combining dwarf failed: Unknown load command 0x32 (50)

make: *** [systemtray] Error 2
```

File: cmd/minishift/cmd/services/systray.go (hunk 2 above)
Comment (LalatenduMohanty) [reply cmd/minishift/cmd/services/systray.go]: I agree with @gbraad , this solution does not scale.

**Comment by LalatenduMohanty** [comment]
> Also `make systemtray` is failing for me on macOS

Started working for me after updating to go version 1.11.2

**Comment by anjannath** [comment]
@gbraad @LalatenduMohanty We'd need two commands one for actually running the service, another to run it as a background process. e.g in the current state of the PR when you run `minishift service systray` you'll launch the tray but it will not start in the background, we launch this command as a process and put it to background from the `minishift services systemtray` command. https://github.com/minishift/minishift/pull/3118/files#diff-55bdb92ce4e1fd01634391ad07105e31R47

For clarity of naming we can append `daemon` to the sub-commands that start service in the background, i.e there will be two commands for the tray `systemtray` and `systemtray-daemon`, the latter is supposed to start the tray in background. (We can come up with better names if needs to be)

Now for the actual structure of the `service(s)` command, we also have the requirement to implement a `kill` command for the services #3062, So it'll be better if we add two sub-command `start` and `stop` to `services` command.

 ```zsh
$ minishift service start systemtray-daemon
$ minishift service stop systemtray-daemon
```

**Comment by gbraad** [comment]
can yo elaborate more about the actual structure and how you are going to fix/solve this? What will the final interaction look like?

I understand the need for two commands, but perhaps in that case, we might have to move the actual interaction with daemon services into a hidden `daemon-services` as namespace?

**Comment by anjannath** [comment]
@gbraad The files inside the `cmd/minishift/cmd/services` are as follows:
```
cmd/minishift/cmd/services/
├── list.go
├── proxy.go
├── services.go
├── sftpd.go
├── start.go
├── stop.go
└── systemtray.go
```
Now there is no cobra command defined to run the services in background, that is handled in `start.go` and likewise for stopping the service is handled in `stop.go`. When adding a new service, one just needs to write the implementation for the service, and can add code to `start.go` to start it in background. I have also added a command `list` to list all the available minishift services.
The cli structure is still the same, e.g to start the tray in background you run the command: `minishift service start systemtray`.

The actual command though, that is used to actually start the tray (without putting it to background) is still the same `minishift service systemtray`

File: cmd/minishift/cmd/services/stop.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	switch service {
+	case minishiftConstants.SystemtrayDaemon:
+		if runtime.GOOS == "linux" {
+			fmt.Println("System tray is not supported in Linux")
+		}
+		minishiftTray := systemtray.NewMinishiftTray(minishiftConfig.AllInstancesConfig)
+		if pid := minishiftTray.GetPID(); pid > 0 {
+			proc, _ := os.FindProcess(pid)
+			proc.Kill()
+			fmt.Printf("Killed process with PID: %d\n", pid)
```
[the remark this answers is not in the document]
Comment (agajdosi) [reply cmd/minishift/cmd/services/stop.go]: @anjannath Ok, so using that binary the behaviour changed, however I find it still problematic:
- `systemtray start` starts the tray, however the command is stuck in execution, when ctrl-c to exit, the tray dissapears.
- when tray is started with `minishift start` and I try to stop it, new tray is started and it is stuck in execution, ctrl-c to exit closes the newly created tray and the old one (from minishift start) persists.

I think we should do a bjn on Monday.

**Comment by LalatenduMohanty** [comment]
Tested on macOS and it seems to workfine.

**Comment by LalatenduMohanty** [comment]
@anjannath @gbraad With the current PR,`minishift services systemtray` is the internal command gets executed when user runs `minishift services start systemtray`. It kind of works for me. However it would be great if the internal command would be `minishift daemon systemtray`. However if the current implementation works for you we can go ahead and merge.

File: cmd/minishift/cmd/daemon/daemon.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+package daemon
+
+import (
+	"github.com/spf13/cobra"
+)
+
+var DaemonCmd = &cobra.Command{
+	Use:   "daemon",
+	Short: "Starts a minishift service daemon",
```
Comment (gbraad) [inline cmd/minishift/cmd/daemon/daemon.go]: Please add this is only to be used by minishift, maybe?

File: cmd/minishift/cmd/daemon/proxy.go [hunk 6]
```diff
@@ -46,7 +46,7 @@ func init() {
 	servicesProxyCmd.Flags().IntVarP(&proxyServerPortFromFlag, proxyServerPortFlag, "p", 3128, "The server port.")
 	servicesProxyCmd.Flags().StringVarP(&proxyUpstreamAddrFromFlag, proxyUpstreamFlag, "u", "", "The upstream proxy address.")
 	servicesProxyCmd.Flags().BoolVarP(&proxyReEncryptFlag, "reencrypt", "r", false, "Re-encrypt traffic")
-	ServicesCmd.AddCommand(servicesProxyCmd)
+	DaemonCmd.AddCommand(servicesProxyCmd)
```
Comment (gbraad) [inline cmd/minishift/cmd/daemon/proxy.go]: `servicesProxyCmd`?

Comment (anjannath) [reply cmd/minishift/cmd/daemon/proxy.go]: Yikes!!... Fixing it! I thought i fixed the names...

File: cmd/minishift/cmd/services/stop.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+			atexit.ExitWithMessage(0, "System tray is not supported in Linux")
+		}
+		minishiftTray := systemtray.NewMinishiftTray(minishiftConfig.AllInstancesConfig)
+		if pid := minishiftTray.GetPID(); pid > 0 {
+			proc, _ := os.FindProcess(pid)
+			proc.Kill()
+			atexit.ExitWithMessage(0, fmt.Sprintf("Killed process with PID: %d\n", pid))
+		}
+	case minishiftConstants.SftpdDaemon:
+		atexit.ExitWithMessage(0, "This is not yet implemented")
```
Comment (gbraad) [inline cmd/minishift/cmd/services/stop.go]: What would it take to support this?

File: cmd/minishift/cmd/services/services.go [hunk 8]
```diff
@@ -21,11 +21,11 @@ import (
 )
 
 var ServicesCmd = &cobra.Command{
-	Use:   "services SUBCOMMAND [flags]",
-	Short: "Manages services used by Minishift.",
-	Long:  `Manages services used by Minishift.`,
+	Use:     "services SUBCOMMAND [flags]",
+	Aliases: []string{"service"},
+	Short:   "Manages services used by Minishift.",
+	Long:    `Manages services used by Minishift.`,
```
[the remark this answers is not in the document]
Comment (anjannath) [reply cmd/minishift/cmd/services/services.go]: To me both of these sentences mean the same thing :)

File: cmd/minishift/cmd/services/stop.go (hunk 7 above)
Comment (anjannath) [reply cmd/minishift/cmd/services/stop.go]: Implemented stop functionality for SFTP daemon.

File: cmd/minishift/cmd/daemon/daemon.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+package daemon
+
+import (
+	"github.com/spf13/cobra"
+)
+
+var DaemonCmd = &cobra.Command{
+	Use:   "daemon",
+	Short: "Starts a minishift service daemon.",
+	Long:  `Starts a minishift service daemon, this is only to be used internally by Minishift`,
```
Comment (LalatenduMohanty) [inline cmd/minishift/cmd/daemon/daemon.go]: `Starts a minishift service daemon, this is only to be used internally by Minishift`. This does not not need a comma.

File: docs/source/using/experimental-features.adoc [hunk 10]
```diff
@@ -295,6 +295,12 @@ the following command:
 $ minishift config set auto-start-tray false
 ----
 
+To start the tray as a daemon process without starting the cluster:
```
Comment (LalatenduMohanty) [inline docs/source/using/experimental-features.adoc]: This should be `To start the tray without starting the cluster:`

Comment (LalatenduMohanty) [reply docs/source/using/experimental-features.adoc]: Also `system tray` is two words. Will fix that while merging the PR.

**Comment by LalatenduMohanty** [comment]
Merged https://github.com/minishift/minishift/commit/ee081a6ad101bad940ab3fe7310f70e329771bef . Thanks @anjannath

**Pull request closed without merging**